src/lib.rs

```diff
@@ -1318,20 +1331,31 @@ impl<T> Clone for ThinVec<T>
 where
     T: Clone,
 {
+    #[inline]
     fn clone(&self) -> ThinVec<T> {
-        let len = self.len();
-        let mut new_vec = ThinVec::<T>::with_capacity(len);
-        let mut data_raw = new_vec.data_raw();
-        for x in self.iter() {
+        #[cold]
+        #[inline(never)]
+        fn clone_non_singleton<T: Clone>(this: &ThinVec<T>) -> ThinVec<T> {
+            let len = this.len();
+            let mut new_vec = ThinVec::<T>::with_capacity(len);
+            let mut data_raw = new_vec.data_raw();
+            for x in this.iter() {
+                unsafe {
+                    ptr::write(data_raw, x.clone());
+                    data_raw = data_raw.add(1);
+                }
+            }
             unsafe {
-                ptr::write(data_raw, x.clone());
-                data_raw = data_raw.add(1);
+                new_vec.set_len(len); // could be the singleton
```

Comment:
not anymore, right?

Reply:
I too made this mistake. `this` is the vector that we have checked for non-singleton-ness, not `new_vec`. If `this` is empty but not the singleton, then `new_vec` will be the singleton.

Reply:
Ah yes, good call (maybe elaborate the comment since that's subtle?)

Reply:
The existing tests did not catch this mistake, but rustc itself crashed when I used `set_let_non_singleton` here. I have now added a new test that will catch this mistake.